File: Hard/AlienDictionary.js
/**
 * @param {string[]} words
 * @return {string}
 */
var alienOrder = function(words) {
    
  const adjList = {};
  for (let word of words)
    for (let c of word)
      adjList[c] = [];
  
  // build adjacency list
  for (let i = 1; i < words.length; i++) {
    let word1 = words[i - 1];
    let word2 = words[i];
    let minLength = Math.min(word1.length, word2.length);
    
    // edge case check for invalid ordering
    if (word1.length > word2.length && word1.startsWith(word2))
        return "";
    
    for (let j = 0; j < minLength; j++) {
      let c1 = word1.charAt(j);
      let c2 = word2.charAt(j);
      if (c1 !== c2) {
        adjList[c2].push(c1); // point edges in opposite direction
        break;
      }
    }
  }
  
  let res = "";
  const visited = {}; // true=visted, false=current path
  let dfs = function(node) {
    if (node in visited)
      return visited[node];
  
    visited[node] = false;
    for(let neighbor of adjList[node]) {
      if (!dfs(neighbor))
        return false;
    }
    res += node;
    visited[node] = true;
    return true;
  }
  
  for (let key in adjList)
    if (!dfs(key))
      return "";
  
  return res;
};

/*
  
  Topological Sort
  1. Loop for every pair of adjacent words and find the first character that differs
  2. Create a directed edge between those two characters and build a graph
  3. The direction of the edges should be reversed so that a post-order traversal returns the right order
  4. Run a DFS that checks for cycles using graph coloring
  
  Graph coloring:
  unvisited = white
  current path = grey
  visited = black

  undefined = white
  false = grey
  true = visited
  
  Time: O(c) - Where c is the total # of characters
  Space: O(1) or O(u + min(u^2,N)) - Since u is fixed at 26
  
  Where u is the total # of unique characters and N is the total number of words
  * There are N - 1 pairs of adjacent words

*/

File: C#/Hard/AlienDictionary.cs
public class Solution
{
    public string AlienOrder(string[] words)
    {
        var result = new List<char>();
        var adjList = new Dictionary<char, List<char>>();
        var visited = new Dictionary<char, bool>(); // true=in current path, false=processed

        // Init the graph for each unique character
        for (int i = 0; i < words.Length; i++)
        {
            foreach (char c in words[i])
            {
                if (!adjList.ContainsKey(c)) adjList.Add(c, new List<char>());
            }
        }

        // Add relationships to the graph
        for (int i = 1; i < words.Length; i++)
        {
            string w1 = words[i - 1];
            string w2 = words[i];
            int len = Math.Min(w1.Length, w2.Length);
            bool samePrefix = true;
            for (int j = 0; j < len; j++)
            {
                char c1 = w1[j];
                char c2 = w2[j];
                if (c1 != c2)
                {    
                    if (!adjList.ContainsKey(c1)) adjList.Add(c1, new List<char>());
                    adjList[c1].Add(c2);
                    samePrefix = false;
                    break;
                }
            }

            // An important edge case - if the prefixes are the same, the smaller word should ALWAYS come first.
            // If the smaller of two matching prefixes doesnt come first, then the order is invalid.
            if (samePrefix && w1.Length > w2.Length) return "";
        }

        // Run topological sort and check for cycles
        bool dfs(char c)
        {
            if (visited.ContainsKey(c)) return visited[c];
            visited.Add(c, true);

            foreach (char neighbor in adjList[c])
            {
                if (dfs(neighbor)) return true;
            }

            result.Add(c);
            visited[c] = false;
            return false;
        }

        // Run topolocal sort on each unvisited piece/subgraph checking for cycles
        foreach (var pair in adjList)
        {
            if (dfs(pair.Key)) return "";
        }

        // Reverse the output string
        result.Reverse();
        return new string(result.ToArray());
    }
}

/*

    The first step to solving this problem is understanding lexographical ordering.

    English letters have the following lexigraphical ordering "abcd..."

    s2 is lexigraphically smaller than s1
    string s1 = "abc"
    string s2 = "aba"

    s1 is lexigraphically smaller than s2
    string s1 = "ab"
    string s2 = "aba"

    s2 is lexigraphically smaller than s1
    string s1 = "ab"
    string s2 = "aaa"

    Solution:
    1. Derive the dependency relationship between letters
    2. Build a directed graph from those dependency relationships
        - letters as nodes
        - relationships as edges
    3. Run a topological sort on the graph to derive a valid ordering (there may be more than one)
        - If a cycle is found, then return an empty string
        - If no cycle is found, return the ordering
    
    Time: O(c)
    Space: O(1)

    Where c is the total number of characters in each word added together

*/
